### ISSUE TITLE
Compiling Grok w/ Pillow (and IIP server)

### ISSUE THREAD

**Issue by calvinbutcher** [GrokImageCompression/grok#7]
Hello,

Having successfully compiled IIP and Loris image servers with OpenJPEG 2.1 (https://github.com/bodleian/iipsrv-openjpeg-docker and https://github.com/bodleian/loris-openjpeg-docker), we're now trying to provide a docker image with your optimised fork of OpenJPEG, Grok 1.0.

As far as I can tell, Pillow (the Python Imaging Library) does not like Grok 1.0, while it will work happily with OpenJPEG 2.1.

I get a Pillow compile error (when attempting to integrate w/ IIP server) as follows:

```
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DHAVE_LIBJPEG -DHAVE_OPENJPEG -DHAVE_LIBZ -DHAVE_LIBTIFF -I/usr/local/include/openjpeg-2.1 -I/tmp/pip-build-0XyZIt/Pillow/libImaging -I/usr/include -I/usr/local/include -I/usr/include/python2.7 -I/usr/include/x86_64-linux-gnu -c libImaging/Jpeg2KDecode.c -o build/temp.linux-x86_64-2.7/libImaging/Jpeg2KDecode.o
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1013:1: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
     OPJ_API void OPJ_CALLCONV opj_cleanup();
     ^
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1662:1: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
     OPJ_API uint32_t OPJ_CALLCONV opj_plugin_get_debug_state();
     ^
    libImaging/Jpeg2KDecode.c: In function ‘j2k_decode_entry’:
    libImaging/Jpeg2KDecode.c:574:5: error: too few arguments to function ‘opj_stream_set_user_data’
         opj_stream_set_user_data(stream, decoder);
         ^
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1125:27: note: declared here
     OPJ_API void OPJ_CALLCONV opj_stream_set_user_data (opj_stream_t* p_stream, void * p_data, opj_stream_free_user_data_fn p_function);
                               ^
    libImaging/Jpeg2KDecode.c:688:35: warning: pointer targets in passing argument 5 of ‘opj_read_tile_header’ differ in signedness [-Wpointer-sign]
                                       &should_continue)) {
                                       ^
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1427:27: note: expected ‘uint32_t *’ but argument is of type ‘OPJ_INT32 *’
     OPJ_API bool OPJ_CALLCONV opj_read_tile_header( opj_codec_t *p_codec,
                               ^
    libImaging/Jpeg2KDecode.c:688:35: warning: pointer targets in passing argument 6 of ‘opj_read_tile_header’ differ in signedness [-Wpointer-sign]
                                       &should_continue)) {
                                       ^
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1427:27: note: expected ‘uint32_t *’ but argument is of type ‘OPJ_INT32 *’
     OPJ_API bool OPJ_CALLCONV opj_read_tile_header( opj_codec_t *p_codec,
                               ^
    libImaging/Jpeg2KDecode.c:688:35: warning: pointer targets in passing argument 7 of ‘opj_read_tile_header’ differ in signedness [-Wpointer-sign]
                                       &should_continue)) {
                                       ^
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1427:27: note: expected ‘uint32_t *’ but argument is of type ‘OPJ_INT32 *’
     OPJ_API bool OPJ_CALLCONV opj_read_tile_header( opj_codec_t *p_codec,
                               ^
    libImaging/Jpeg2KDecode.c:688:35: warning: pointer targets in passing argument 8 of ‘opj_read_tile_header’ differ in signedness [-Wpointer-sign]
                                       &should_continue)) {
                                       ^
    In file included from libImaging/Jpeg2K.h:11:0,
                     from libImaging/Jpeg2KDecode.c:21:
    /usr/local/include/openjpeg-2.1/openjpeg.h:1427:27: note: expected ‘uint32_t *’ but argument is of type ‘OPJ_INT32 *’
     OPJ_API bool OPJ_CALLCONV opj_read_tile_header( opj_codec_t *p_codec,
                               ^
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

```

When compiling Pillow with Loris integration I get:

`error: unknown type name 'OPJ_CINEMA_MODE'`

OPJ_CINEMA_MODE is a deprecated feature as listed here:

> "typedef enum CINEMA_MODE OPJ_CINEMA_MODE
> 
> DEPRECATED: use RSIZ, OPJ_PROFILE_* and OPJ_EXTENSION_* instead Digital cinema operation mode."
> 

from http://www.openjpeg.org/doxygen/openjpeg_8h.html

So my question is: is Grok 1.0 perhaps based on a pre-2.1 version of OpenJPEG that will not compile with Pillow/PIL? Or am I doing something wrong?

My sincere apologies if I'm making a simple error here. I'm a bit of a newbie when it comes to this sort of thing.

Best regards,
Calvin

**Comment by boxerab** [comment]
Hi Calvin,

Thanks for your interest in the library. I would advise using the latest version in master, not the 1.0 tag.  I will help you get this working.

Cheers,
Aaron

**Comment by calvinbutcher** [comment]
Hi Aaron,

Thanks for the swift response.

Unfortunately I've already tried the master.

Best,
C

**Comment by calvinbutcher** [comment]
This is the account for the docker image I'm trying to build (rather than just repeating everything here):

https://github.com/bodleian/iipsrv-grok-docker

In that particular file it does checkout the 1.0 tag, but I've since tried a manual build on a fresh Ubuntu 14.04 VM with the master branch but to no avail.

**Comment by boxerab** [comment]
Thanks. Let me investigate.

**Comment by boxerab** [comment]
Grok is a C++11 project, by the way. Also, IIP server doesn't officially support openjpeg.  Did you patch IIP server to add support ?

Also, in the output from gcc, I see a lot of warnings, then gcc errors out. Are you treating warnings as errors?

**Comment by calvinbutcher** [comment]
IIP server compiles with stweil's openjpeg implementation. Please see https://github.com/moravianlibrary/iipsrv-openjpeg/issues/2.

A "pip install Pillow" (as per the process in the above docker file) results in a successful install using OpenJPEG 2.1.

With Grok 1.0 the "pip install Pillow"  results in the warnings given above; gcc then errors out, as you mention, and I receive this final error from - presumably - pip:

`Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-hHQmAG/Pillow/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-1MyvmL-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-hHQmAG/Pillow/`

I'm afraid I'm unfamiliar with C++ projects. Are you saying I'm using the wrong compiler/version (i.e. gcc)?

**Comment by calvinbutcher** [comment]
Did you modify my comment with "Unfortunately, I am unable to get the docker file, following the instructions on the github project." ?

There is no dockerfile (as per the instructions) yet as it will not compile, i.e. there is no "latest" tag made available from a successful compile. However, the file `dockerfile` contains all of the installation steps. I am currently going through these instructions manually on a VM, trying to test and debug.

**Comment by calvinbutcher** [comment]
Thanks for your time, by the way :)

**Comment by calvinbutcher** [comment]
My apologies, I'm getting mixed up here. The Pillow install is, I believe, for the benefit of the IIIF validator. I'm not so bothered about that.

IIP server does appear to configure and compile with Grok fine (it declares it has openjpeg support). However, when run with a jp2, e.g.:

`http://localhost/fcgi-bin/iipsrv.fcgi?IIIF=67352ccc-d1b0-11e1-89ae-279075081939.jp2/full/full/0/default.jpg`

I get a 500, the error for which is in the apache log is:

```
[Thu May 26 12:38:41.391072 2016] [fcgid:warn] [pid 30760:tid 140100398696192] [client 127.0.0.1:37668] mod_fcgid: error reading data, FastCGI server closed connection
[Thu May 26 12:38:41.391159 2016] [core:error] [pid 30760:tid 140100398696192] [client 127.0.0.1:37668] End of script output before headers: iipsrv.fcgi
[Thu May 26 12:38:46.637207 2016] [fcgid:error] [pid 30758:tid 140100664620928] mod_fcgid: process /var/www/localhost/fcgi-bin/iipsrv.fcgi(30851) exit(communication error), get signal 6, possible coredump generated
```

But the IIIF parameter call works fine for a TIF, and when compiled with OpenJPEG 2.1 (not Grok) it works fine with a JP2.

My sincere apologies if I'm muddying the waters, I'm still getting my head around these builds.

**Comment by boxerab** [comment]
Calvin,

Let's isolate the problem. Let's forget about docker and Pillow.

Can you test IIPServer with fcgi stand alone using spawn-fcgi ?  

http://iipimage.sourceforge.net/documentation/server/

Then you might get more info about the core dump.

Also, there is a pull request on github for openjpeg support:

https://github.com/ruven/iipsrv/pull/61

Please give this an upvote so it gets merged soon :)

Thanks,
Aaron

**Comment by calvinbutcher** [comment]
Agreed. I've upvoted. I was just in the process of trying to generate a core dump but I've nothing so far. Now I just get:

```
[Thu May 26 13:00:13.049824 2016] [fcgid:warn] [pid 17444:tid 139753402312448] [client 127.0.0.1:37810] mod_fcgid: error reading data, FastCGI server closed connection
[Thu May 26 13:00:13.049937 2016] [core:error] [pid 17444:tid 139753402312448] [client 127.0.0.1:37810] End of script output before headers: iipsrv.fcgi
```

.. with no 'possible coredump' entry in the error log, as well as no coredump in my designated (and correctly permissioned) CoreDumpDirectory.

As regards spawn-fcgi, are you requesting I set this up with nginx or simply spawn the fcgi processes from the CLI under my current Apache deployment. Will this help me get a core dump?

**Comment by boxerab** [comment]
Hi Calvin.  First things first: can you build Grok on your system, and decompress an image?

$ opj_decompress -i FOO.jp2 -o BAR.tif -OutFor TIF

**Comment by calvinbutcher** [comment]
I get the following. It seems jp2 isn't available as a format. It gives the error twice.

```
grok@grok-VirtualBox:~/openjpeg/bin$ ./opj_compress -i /var/www/localhost/images/67352ccc-d1b0-11e1-89ae-279075081939.jp2 -o out.tif -OutFor TIF
[ERROR] Unknown input file format: /var/www/localhost/images/67352ccc-d1b0-11e1-89ae-279075081939.jp2 
        Known file formats are *.pnm, *.pgm, *.ppm, *.pgx, *png, *.bmp, *.tif, *.raw or *.tga
[ERROR] Unknown input file format: /var/www/localhost/images/67352ccc-d1b0-11e1-89ae-279075081939.jp2 
        Known file formats are *.pnm, *.pgm, *.ppm, *.pgx, *png, *.bmp, *.tif, *.raw or *.tga
```

**Comment by calvinbutcher** [comment]
I'm still getting the same error (though now only once). IIPsrv configured and installed correctly with Grok, declaring it had openjpeg functionality. There were no compilation issues with Grok and I'm using the master branch.

The jp2 image is from here: http://iiif-test.stanford.edu/67352ccc-d1b0-11e1-89ae-279075081939.jp2 and works fine with Kakadu and OpenJPEG.

**Comment by boxerab** [comment]
The command was opj_decompress, not opj_compress, but it looks like grok is installed.

One thing - I think you need to build as a shared library. This is an option in CMake. Also, make sure build type is Release, for maximum performance.

**Comment by boxerab** [comment]
So, once you build grok in this way, and build IIPSrv, tell me what happens :)

**Comment by boxerab** [comment]
I would really like to help get this working

**Comment by calvinbutcher** [comment]
OK great, thank you. I'll try that. Yes, it would be really nice to get it working.

**Comment by calvinbutcher** [comment]
OK, now I get:

```
grok@grok-VirtualBox:~/grokimagecomp/bin$ ./opj_decompress -i /var/www/localhost/images/67352ccc-d1b0-11e1-89ae-279075081939.jp2 -o out.tif -OutFor TIF
Plugin ./libopenjp2_plugin.so successfully loaded
Full library path: ./libopenjp2_plugin.so
Plugin Loaded and Initialized! 
Plugin ./libopenjp2_plugin.so successfully initialized 

[INFO] Start to read j2k main header (111).
[INFO] Main header has been correctly decoded.
[INFO] No decoded area parameters, set the decoded area to the whole image
[INFO] Header of tile 1 / 1 has been read.
[INFO] Tile 1/1 has been decoded.
[INFO] Stream reached its end !
[INFO] Generated Outfile out.tif
decode time: 227 ms 
```

Which is good. However, I get the same error in apache:
```

[Fri May 27 15:00:31.254299 2016] [fcgid:warn] [pid 30410:tid 139781290006272] [client 127.0.0.1:60876] mod_fcgid: error reading data, FastCGI server closed connection
[Fri May 27 15:00:31.254513 2016] [core:error] [pid 30410:tid 139781290006272] [client 127.0.0.1:60876] End of script output before headers: iipsrv.fcgi
[Fri May 27 15:00:33.257091 2016] [fcgid:error] [pid 30409:tid 139781373872000] mod_fcgid: process /var/www/localhost/fcgi-bin/iipsrv.fcgi(30554) exit(communication error), get signal 6, possible coredump generated
```

IIPsrv compiled with Grok without any complaints.

**Comment by boxerab** [comment]
Progress :)  You need to run this with gdb, I think. Can you do this with apache? Or, perhaps you need to run as standalone FCGI ?  Then you can get debug info.

**Comment by boxerab** [comment]
There is info on fcgi standalone on the IIPsrv github site.

**Comment by boxerab** [comment]
If you figure out standalone, I can try on my side to debug.

**Comment by calvinbutcher** [comment]
Not sure, I've not used it before. I'll look into it and get back to you asap. Thanks again.

**Comment by calvinbutcher** [comment]
OK, well I've recompiled IIPsrv with the relevant CFLAGs (though it still doesn't find the debugging symbols). I've run `gdb usr/sbin/apache2` and `run -X` (https://httpd.apache.org/dev/debugging.html) but get nothing when the 500 occurs. But I'll start again next Tuesday (it's a bank holiday weekend here). This is the first time I've done this so I'm probably missing something.

```
grok@grok-VirtualBox:/home/grok/iip$ sudo -E gdb /usr/sbin/apache2
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/sbin/apache2...(no debugging symbols found)...done.
(gdb) run -X
Starting program: /usr/sbin/apache2 -X
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 127.0.1.1. Set the 'ServerName' directive globally to suppress this message
[New Thread 0x7ffff37e6700 (LWP 4006)]
[New Thread 0x7ffff2fe5700 (LWP 4007)]
[New Thread 0x7ffff27e4700 (LWP 4008)]
[New Thread 0x7ffff1fe3700 (LWP 4009)]
[New Thread 0x7ffff17e2700 (LWP 4010)]
[New Thread 0x7ffff0fe1700 (LWP 4011)]
[New Thread 0x7fffebfff700 (LWP 4012)]
[New Thread 0x7fffeb7fe700 (LWP 4013)]
[New Thread 0x7fffeaffd700 (LWP 4014)]
[New Thread 0x7fffea7fc700 (LWP 4015)]
[New Thread 0x7fffe9ffb700 (LWP 4016)]
[New Thread 0x7fffe97fa700 (LWP 4017)]
[New Thread 0x7fffe8ff9700 (LWP 4018)]
[New Thread 0x7fffe87f8700 (LWP 4019)]
[New Thread 0x7fffe7ff7700 (LWP 4020)]
[New Thread 0x7fffe77f6700 (LWP 4021)]
[New Thread 0x7fffe6ff5700 (LWP 4022)]
[New Thread 0x7fffe67f4700 (LWP 4023)]
[New Thread 0x7fffe5ff3700 (LWP 4024)]
[New Thread 0x7fffe57f2700 (LWP 4025)]
[New Thread 0x7fffe4ff1700 (LWP 4026)]
[New Thread 0x7fffe47f0700 (LWP 4027)]
[New Thread 0x7fffe3fef700 (LWP 4028)]
[New Thread 0x7fffe37ee700 (LWP 4029)]
[New Thread 0x7fffe2fed700 (LWP 4030)]
[New Thread 0x7fffe27ec700 (LWP 4031)]
[New Thread 0x7fffe1feb700 (LWP 4032)]
[Thread 0x7ffff37e6700 (LWP 4006) exited]
```

**Comment by boxerab** [comment]
I think the simplest approach is to skip apache and run fcgi
standalone.  Lets discuss next week.

**Comment by calvinbutcher** [comment]
I looked into this but couldn't find any clear instructions/documentation. 

Running it standalone seemed to have something to do with a Java "JIIViewer" application that I couldn't find anywhere. 

I'm going to start a fresh build now.

**Comment by boxerab** [comment]
Hi Calvin,
Hope you had a good weekend.  Here is an idea: why don't you create a docker image of an iipsrv + openjpeg install, with iipsrv running as stand alone at address 127.0.0.1 on port 9000 ? Then I can download the docker image and try to get this running with Grok.

Also, you may be interested in this thread

https://github.com/ruven/iipsrv/pull/61#issuecomment-222381601

from Stefan Weil, at the University of Mannheim Library.

**Comment by boxerab** [comment]
I mean, running iipsrv from the command line:

https://github.com/ruven/iipsrv

Look at the bottom of the readme for instructions.

**Comment by calvinbutcher** [comment]
Hmm, all very interesting and sounding very promising!

So I'm getting mixed up between "standalone" and command line? Anyway, regarding a docker image running iipsrv with openjpeg (not Grok?), you could download this (instructions in the README):

https://github.com/bodleian/iipsrv-openjpeg-docker

... which is a stable build (see https://hub.docker.com/r/bdlss/iipsrv-openjpeg-docker/builds/)

And simply run:

`./iipsrv.fcgi --bind 192.168.0.1:9000`

from the `/var/www/localhost/fcgi-bin` directory.

But was it iipsrv w/ openjpeg that you wanted? I'm just re-trying gdb with new my build...

**Comment by calvinbutcher** [comment]
Yes, I had a lovely weekend thank you. Did a bit of cycling around Richmond park, had a pint at the Roebuck on Richmond Hill in some rare sunshine.

**Comment by calvinbutcher** [comment]
Updated my last comment

**Comment by boxerab** [comment]
Cycling in the sun and drinking beer : sounds awesome.

I will try the docker image : does Apache need to be told not to spawn its own fastcgi server ?

**Comment by calvinbutcher** [comment]
It was! And you? 

Yes, sorry: you'll need to do a `sudo service apache2 stop` or equivalent thereof.

**Comment by calvinbutcher** [comment]
OK, getting nowhere with gdb and apache. But I'm not certain I'm compiling iipsrv or grok correctly (I'm compilng with shared libs and type debug, etc., as well as -g and setting the CFLAGs in the make). I'm not  familiar with compiling C.

Now that I know by "standalone" you mean CLI, I'll try this direction.

**Comment by boxerab** [comment]
My weekend was good thanks : quality time with the family :)  

I am glad you're joining the iipsrv thread on openjpeg and grok : stefan will be very helpful.

**Comment by calvinbutcher** [comment]
It would appear Stefan has already been successful compiling iipsrv and grok, so fingers crossed..

**Comment by calvinbutcher** [comment]
@boxerab In case your interested, Loris (https://github.com/loris-imageserver/loris) build fails when it tries to install a Pillow >= 2.4.0 dependency (it gets 3.2).

`libImaging/Jpeg2KDecode.c: In function 'j2k_decode_entry': libImaging/Jpeg2KDecode.c:574:5: error: too few arguments to function 'opj_stream_set_user_data' opj_stream_set_user_data(stream, decoder); ^`

Results in:

`
[0mSingle threaded build, not installing mp_compile:1 processes [91merror: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 1[0m[91m [0mRemoving intermediate container 62dddf6d93ca The command '/bin/sh -c ./setup.py install' returned a non-zero code: 1`

I'm looking into it but I'm not really sure I can help from here :( Sorry, I've come back here because I thought it perhaps wasn't relevant to be talking about loris on an iipsrv issue thread.

**Comment by boxerab** [comment]
Hi Calvin,  try looking for the method opj_stream_set_user_data in Grok/openjpeg.h header,
and see what the signature is.  This will be easy to fix :)

**Comment by boxerab** [comment]
I may have changed it and now it is not compatible with openjpeg

**Comment by boxerab** [comment]
you can run around with the number one hand signal when you get this working :)

**Comment by calvinbutcher** [comment]
Cheers! I'll try!

**Comment by calvinbutcher** [comment]
OK, the Grok opj_stream_set_user_data signatures are the same as the openjpeg ones. No difference at all. 

However, in Pillow, I found this:

```
    /* OpenJPEG 2.0 doesn't have OPJ_VERSION_MAJOR */
#ifndef OPJ_VERSION_MAJOR
    opj_stream_set_user_data(stream, decoder);
#else
    opj_stream_set_user_data(stream, decoder, NULL);
```
https://github.com/python-pillow/Pillow/search?utf8=%E2%9C%93&q=opj_stream_set_user_data

But as far as I can tell OPJ_VERSION_MAJOR is set to 2 in the Grok/CMakeLists.txt. However, the error suggests it is _not_ defined:

`error: too few arguments to function 'opj_stream_set_user_data' opj_stream_set_user_data(stream, decoder);`

If Pillow thought OPJ_VER.. was defined, it should have appended a NULL into the third parameter. So I'm a little confused. I think this goes a bit (but perhaps not much) further than the signatures being wrong. 

I'll continue looking into it but thought I'd give you a heads up in case I'm heading down a long, deep, dark rabbit hole here.. Perhaps I should raise this as a Pillow issue? Thanks for your time and my apologies if I'm distracting you from more important work.

**Comment by boxerab** [comment]
Don't worry, we will get to the bottom of this :)  

Can you try to simply #define OPJ_VERSION_MAJOR in this file, to force Pillow to use the correct signature?

**Comment by boxerab** [comment]
Do you have another install of openjpeg on the system, perhaps the headers are being read from somewhere else? A 2.0 install, perhaps.

**Comment by calvinbutcher** [comment]
Pillow is installed from PyPi via a pip install during Loris' setup.py, but perhaps I could pre-compile it and set OPJ_VER as you suggest (or actually I could just remove the condition and force the correct function call).

I don't have another install of openjpeg on the system, no. It's only downloading the Grok source and compiling.

I'll try the first option early tomorrow, thanks!

**Comment by boxerab** [comment]
Can you open `/usr/local/include/openjpeg2.1\opj_config.h` and verify that the version in that file
is 2.1.1 ?

**Comment by boxerab** [comment]
And yes, try hacking the code to force the correct compilation, and then worry about why it is getting the version wrong.

**Comment by calvinbutcher** [comment]
Looking into it now. In the meantime, could I get that release tag? I'd just like to make stable what I can as I believe Stefan would like to use these docker images for a conference. Cheers!

**Comment by calvinbutcher** [comment]
`/usr/local/include/openjpeg2.1/opj_config.h` gives me:

```
#define OPJ_VERSION_MAJOR 2
#define OPJ_VERSION_MINOR 1
#define OPJ_VERSION_BUILD 1
```

**Comment by calvinbutcher** [comment]
OK, so now I get:

```
    libImaging/Jpeg2KEncode.c: In function ‘j2k_set_cinema_params’:
    libImaging/Jpeg2KEncode.c:193:15: error: ‘opj_cparameters_t’ has no member named ‘cp_cinema’
         if (params->cp_cinema == OPJ_CINEMA4K_24) {
                   ^
    libImaging/Jpeg2KEncode.c: In function ‘j2k_encode_entry’:
    libImaging/Jpeg2KEncode.c:271:5: warning: large integer implicitly truncated to unsigned type [-Woverflow]
         unsigned _SIZE__MAX = SIZE_MAX;
         ^
    libImaging/Jpeg2KEncode.c:426:11: error: ‘opj_cparameters_t’ has no member named ‘cp_cinema’
         params.cp_cinema = context->cinema_mode;
               ^
    libImaging/Jpeg2KEncode.c:428:19: error: ‘opj_cparameters_t’ has no member named ‘cp_cinema’
         switch (params.cp_cinema) {
                       ^
    libImaging/Jpeg2KEncode.c:430:15: error: ‘opj_cparameters_t’ has no member named ‘cp_rsiz’
             params.cp_rsiz = OPJ_STD_RSIZ;
                   ^
    libImaging/Jpeg2KEncode.c:434:15: error: ‘opj_cparameters_t’ has no member named ‘cp_rsiz’
             params.cp_rsiz = OPJ_CINEMA2K;
                   ^
    libImaging/Jpeg2KEncode.c:439:15: error: ‘opj_cparameters_t’ has no member named ‘cp_rsiz’
             params.cp_rsiz = OPJ_CINEMA4K;
                   ^
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```

Looking at the code I see:

```
    /* These settings have been copied from opj_compress in the OpenJPEG
       sources. */
```

So looking at `opj_compress` in the Grok source I find all the given parameters but no `cp_cinema`. I'm guessing this is some new 4K thing, so I'm going to try installing an earlier version of Pillow, Loris only requires <=2.4. I think this is probably the issue.

**Comment by boxerab** [comment]
Good morning.  cp_cinema is deprecated in openjpeg, so I removed it in Grok.  I can add it back in if it makes it easier to compile.

**Comment by boxerab** [comment]
btw, I am making a new tag. What would you like me to name it: 2.1.1 ?

**Comment by boxerab** [comment]
btw, how did you solve the version issue?

**Comment by calvinbutcher** [comment]
Hi, the Pillow 2.4 attempt fails to find JPEG2K support, so just another rabbit hole there.

Has cp_cinema been removed from OpenJPEG 2.1? If so, it shouldn't be causing this issue as Loris compiles fine with OpenJPEG 2.1..!?

Thanks, whatever you like, really. Stefan just gave me a "openjpeg-[date]" tag, so if you don't want it to mess with your version numbers perhaps something simliar would be better for you?

**Comment by calvinbutcher** [comment]
I just removed the #ifndef entirely and made a 3 parameter call with NULL.

**Comment by boxerab** [comment]
okey dokey, I have released version 2.1.1, with tag  `v2.1.1`

**Comment by calvinbutcher** [comment]
Thank you, I'll check those out in the docker images and tell Stefan.

**Comment by boxerab** [comment]
cp_cinema is still in openjpeg 2.1, but deprecated. I can add it back in to grok, but it won't do anything :)

**Comment by boxerab** [comment]
let's find the easiest path to getting your docker image working

**Comment by calvinbutcher** [comment]
I guess iipsrv works because it doesn't rely on Pillow in any way. Putting cp_cinema back in should help it compile, if there aren't any further issues? I could then raise it as an issue on the Pillow account: something along the lines of "param cp_cinema is deprecated in openjpeg >=2.1"...

**Comment by boxerab** [comment]
alrighty, I have added the cp_cinema variable back in, and pushed the v2.1.1 tag to point to the latest commit.

**Comment by calvinbutcher** [comment]
OK, great. I'll give that a try now. Fingers crossed!

**Comment by boxerab** [comment]
fingers crossed. It would be nice to understand why the correct version is not being detected.

**Comment by boxerab** [comment]
but leave that for later....

**Comment by calvinbutcher** [comment]
Yes, it's strange. But I'll check it'll fully compile first, then that can be looked at.

**Comment by boxerab** [comment]
wait, I need to add cp_rsiz as well :(

**Comment by boxerab** [comment]
done, you can try it now.

**Comment by calvinbutcher** [comment]
Well, although the unit tests all fail I do get:
```

--------------------------------------------------------------------
Pillow 3.2.0 TEST SUMMARY 
--------------------------------------------------------------------
Python modules loaded from /home/grok/Pillow/Pillow-3.2.0/PIL
Binary modules loaded from /home/grok/Pillow/Pillow-3.2.0/PIL
--------------------------------------------------------------------
--- PIL CORE support ok
*** TKINTER support not installed
--- FREETYPE2 support ok
--- LITTLECMS2 support ok
*** WEBP support not installed
--- JPEG support ok
--- OPENJPEG (JPEG2000) support ok
--- ZLIB (PNG/ZIP) support ok
--- LIBTIFF support ok
```

So it's a good start! I'm going to install Loris now..

**Comment by calvinbutcher** [comment]
Loris installs...

**Comment by calvinbutcher** [comment]
it works :)

**Comment by calvinbutcher** [comment]
The only remaining question being, why isn't it picking up the correct version and thus making the 3 param call?

**Comment by calvinbutcher** [comment]
Have a nice weekend!

**Comment by boxerab** [comment]
Awesome!! Nice work.  We can try figuring out the version issue next week.
Have a great weekend!

**Comment by calvinbutcher** [comment]
OK, for now I'm going to template in a revised copy of jpeg2kDecode/Encode.c before compiling pillow. Clunky but at least the docker image will work until this is resolved.

**Comment by boxerab** [comment]
Good plan.

**Comment by calvinbutcher** [comment]
Just an FYI. When installing manually and ignoring a non-zero exit status for Pillow self tests, Loris works with Grok (given the changes made to Pillow).

Under a docker image build, however, I now get the following error:

` libImaging/Jpeg2KDecode.c: In function 'j2k_decode_entry': libImaging/Jpeg2KDecode.c:574:38: error: 'encoder' undeclared (first use in this function) opj_stream_set_user_data(stream, encoder, NULL);`

Strangely, this error doesn't occur when installing manually. I'll continue trying a few things but I just thought I'd update you.

**Comment by boxerab** [comment]
Thanks for the update. We really need to get to the bottom of these errors, starting with the version issue. Not sure how to proceed, though.

**Comment by calvinbutcher** [comment]
OK, the failing Pillow selftests seem to be related to this issue, marked as a bug (in Pillow).

https://github.com/python-pillow/Pillow/issues/1204

It seems to have been left hanging since last year, though a partial fix is said to have been merged.

**Comment by boxerab** [comment]
wait, are you running on a big endian machine ?

**Comment by calvinbutcher** [comment]
My sincere apologies, that last error was my bad. I've deleted the thread entry to avoid confusion.

**Comment by calvinbutcher** [comment]
I'm running on a docker image of ubuntu 14.04. I'll see how it goes now I've removed my stupid mistake!

**Comment by boxerab** [comment]
No worries !

**Comment by calvinbutcher** [comment]
Yes, so that inital error was me being a clutz. The docker image build now fails on these selftests, however. It works if they are ignored (i.e. it is built manually) but docker suspends the build. It seems to be a similar issue as that given at https://github.com/python-pillow/Pillow/issues/1204. However, there is also a ".libpng warning: Invalid color type/bit depth combination in IHDR
libpng error: Invalid IHDR data". So perhaps that is kicking it off, I'll look into it.

Sorry, I'm really just bumbling my way through all this. I'm not really familiar with github, let alone C++. Thanks for your patience.

```
Installed /home/grok/pillow
Processing dependencies for Pillow==3.2.0
Finished processing dependencies for Pillow==3.2.0
python selftest.py
--------------------------------------------------------------------
Pillow 3.2.0 TEST SUMMARY 
--------------------------------------------------------------------
Python modules loaded from /home/grok/pillow/PIL
Binary modules loaded from /home/grok/pillow/PIL
--------------------------------------------------------------------
--- PIL CORE support ok
*** TKINTER support not installed
--- FREETYPE2 support ok
--- LITTLECMS2 support ok
*** WEBP support not installed
--- JPEG support ok
--- OPENJPEG (JPEG2000) support ok
--- ZLIB (PNG/ZIP) support ok
--- LIBTIFF support ok
--------------------------------------------------------------------
Running selftest:
--- 57 tests passed.
nosetests Tests/test_*.py
.....................SSSSSSS........................................................................SS...................S...........................S............S..FFFF.F.FFFFFFFF...FF.............................................S..........................................................................................................SSSSSSSSSSSSS...............................S.................................SSSSS.......................................................libpng warning: Invalid color type/bit depth combination in IHDR
libpng error: Invalid IHDR data
.S.................................................S......................................................S..............................................................S...SSSSSS........SSSSSS..................SSSSSS..........
======================================================================
FAIL: TestFileJpeg2k.test_16bit_j2k_roundtrips
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 165, in test_16bit_j2k_roundtrips
    self.assert_image_equal(im, j2k)
  File "/home/grok/pillow/Tests/helper.py", line 76, in assert_image_equal
    self.fail(msg or "got different content")
AssertionError: got different content

======================================================================
FAIL: TestFileJpeg2k.test_16bit_jp2_roundtrips
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 171, in test_16bit_jp2_roundtrips
    self.assert_image_equal(im, jp2)
  File "/home/grok/pillow/Tests/helper.py", line 76, in assert_image_equal
    self.fail(msg or "got different content")
AssertionError: got different content

======================================================================
FAIL: TestFileJpeg2k.test_16bit_monchrome_j2k_like_tiff
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 159, in test_16bit_monchrome_j2k_like_tiff
    self.assert_image_similar(j2k, tiff_16bit, 1e-3)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 218.6460 > epsilon 0.0010

======================================================================
FAIL: TestFileJpeg2k.test_16bit_monchrome_jp2_like_tiff
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 153, in test_16bit_monchrome_jp2_like_tiff
    self.assert_image_similar(jp2, tiff_16bit, 1e-3)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 244.6838 > epsilon 0.0010

======================================================================
FAIL: TestFileJpeg2k.test_bytesio
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 54, in test_bytesio
    self.assert_image_similar(im, test_card, 1.0e-3)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 245.8860 > epsilon 0.0010

======================================================================
FAIL: TestFileJpeg2k.test_irreversible_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 91, in test_irreversible_rt
    self.assert_image_similar(im, test_card, 2.0)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 246.5775 > epsilon 2.0000

======================================================================
FAIL: TestFileJpeg2k.test_layers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 117, in test_layers
    self.assert_image_similar(im, test_card, 13)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 309.1326 > epsilon 13.0000

======================================================================
FAIL: TestFileJpeg2k.test_lossless
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 64, in test_lossless
    self.assert_image_similar(im, test_card, 1.0e-3)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 245.8860 > epsilon 0.0010

======================================================================
FAIL: TestFileJpeg2k.test_lossless_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 73, in test_lossless_rt
    self.assert_image_equal(im, test_card)
  File "/home/grok/pillow/Tests/helper.py", line 76, in assert_image_equal
    self.fail(msg or "got different content")
AssertionError: got different content

======================================================================
FAIL: TestFileJpeg2k.test_lossy_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 77, in test_lossy_rt
    self.assert_image_similar(im, test_card, 2.0)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 299.2443 > epsilon 2.0000

======================================================================
FAIL: TestFileJpeg2k.test_lossy_tiled
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 69, in test_lossy_tiled
    self.assert_image_similar(im, test_card, 2.0)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 407.7864 > epsilon 2.0000

======================================================================
FAIL: TestFileJpeg2k.test_prog_qual_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 96, in test_prog_qual_rt
    self.assert_image_similar(im, test_card, 2.0)
  File "/home/grok/pillow/Tests/helper.py", line 100, in assert_image_similar
    ave_diff, epsilon))
AssertionError:  average pixel value difference 315.7644 > epsilon 2.0000

======================================================================
FAIL: TestFileJpeg2k.test_prog_res_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 100, in test_prog_res_rt
    self.assert_image_equal(im, test_card)
  File "/home/grok/pillow/Tests/helper.py", line 76, in assert_image_equal
    self.fail(msg or "got different content")
AssertionError: got different content

======================================================================
FAIL: TestFileJpeg2k.test_tiled_offset_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 87, in test_tiled_offset_rt
    self.assert_image_equal(im, test_card)
  File "/home/grok/pillow/Tests/helper.py", line 76, in assert_image_equal
    self.fail(msg or "got different content")
AssertionError: got different content

======================================================================
FAIL: TestFileJpeg2k.test_tiled_rt
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/grok/pillow/Tests/test_file_jpeg2k.py", line 81, in test_tiled_rt
    self.assert_image_equal(im, test_card)
  File "/home/grok/pillow/Tests/helper.py", line 76, in assert_image_equal
    self.fail(msg or "got different content")
AssertionError: got different content

----------------------------------------------------------------------
Ran 702 tests in 15.494s

FAILED (SKIP=54, failures=15)
make: *** [release-test] Error 1
```

**Comment by calvinbutcher** [comment]
So these pillow test images work:

`16bit.cropped.jp2  rgb_trns_ycbc.jp2  test-card-lossless.jp2  test-card-lossy-tiled.jp2`

and these cause a 500:

`unbound_variable.jp2   16bit.cropped.j2k  rgb_trns_ycbc.j2k`

Not sure if that's right or wrong. Maybe pillow is out of step with openjpeg 2.1.

**Comment by boxerab** [comment]
Can you point me to where these test images are? I can test them directly with Grok.

**Comment by boxerab** [comment]
You might want to post these issues on [stack overflow](http://stackoverflow.com/)
You should get plenty of assistance there.

**Comment by boxerab** [comment]
I found the test images : the first one has an error in it, but the others should decode fine.

**Comment by calvinbutcher** [comment]
Further investigation has lead me to this issue (though it is two years old, see bold):

> openjpeg-2.1.3 does not exist. But openjpeg-2.1.0 and openjpeg-2.0.1 do, and there are some issue with those versions:
> 
>    - For both openjpeg-2.1.0 and openjpeg-2.0.1, **the tests fail**. The upstream commit which introduced the failure is https://code.google.com/p/openjpeg/source/detail?r=2317 . I'm not sure whether the commit is bad or the test images. Would be nice if the author of the jpeg2k plugin, @al45tair, could give it a look.
>
>    - For openjpeg-2.1.0, they moved the headers to `/usr/include/openjpeg-2.1`, and introduced some API changes. setup.py needs to be modified for 2.1.0, and **the changes for the new API are easy** (`opj_stream_set_user_data(stream, decoder)` -> `opj_stream_set_user_data(stream, decoder, NULL)`).

https://github.com/python-pillow/Pillow/issues/623

**Comment by calvinbutcher** [comment]
So looks to me like Pillow is just a bit behind with regard to openjpeg, and it's already been raised some time ago.

**Comment by boxerab** [comment]
I suspected as much :)  You could be the openjpeg evangelist for Pillow :)
I am sure if you bring this up and badger them a little, they will fix it, or perhaps guide you on fixing it yourself.

**Comment by wiredfool** [comment]
Your issue doesn't sound related to any of the issues you've found in the pillow project. 

It sounds like Grok is a fork of OpenJpeg, and isn't strictly compatible with either 2.0.0 or 2.1.0. If it's going to work with Pillow, one of them is going to need to be modified. The Jpeg2k support hasn't had a whole lot of attention since it was initially merged, so I wouldn't be surprised that there are places that need updating. 

Re: Versioning, there's some version detection in setup.py to determine the best version of openjpeg to use. You may want to check that it will detect Grok's header setup. There's a `--debug` flag for build_ext that will trace what headers and libraries are found. 

There's one patch outstanding that's likely to be merged: https://github.com/python-pillow/Pillow/pull/1934 . I don't think that it significantly changes the api that we're using on openjpeg, but it's worth a look.

**Comment by calvinbutcher** [comment]
@wiredfool thanks, that may explain why pillow isn't finding the correct openjpeg version. I'm going to have to look this `build_ext` business up.

That alone would resolve a chunk of the problem. As regards the tests, they do pass with openjpeg 2.1, which builds without error against Loris.

**Comment by wiredfool** [comment]
The command is: `python setup.py build_ext --debug`, followed by `python setup.py install`.  The `--debug` flag is new enough that it's only in master, not a release yet. So you're going to have to pull from the git repo rather than pypi.

**Comment by calvinbutcher** [comment]
Great! Thank you. I'm pulling from git anyway.

**Comment by calvinbutcher** [comment]
OK, using master. From `PIL SETUP SUMMARY` I'm getting:

`--- OPENJPEG (JPEG2000) support available (2.1)`

Which would suggest to me it's finding the [correct] version. So there must be a disconnect between this version method and the one used to stipulate the 2 or 3 paramater call to `opj_stream_user_data` in `Jpeg2KDecode.c` and `Jpeg2KEncode.c`.

After this the debug info (if I'm looking in the right place) only really gives me the following. Can't seem to find any other references to 2K/2000 or openjpeg:

```
copying PIL/Jpeg2KImagePlugin.py -> build/lib.linux-x86_64-2.7/PIL
...
copying build/lib.linux-x86_64-2.7/PIL/Jpeg2KImagePlugin.py -> build/bdist.linux-x86_64/egg/PIL
...
byte-compiling build/bdist.linux-x86_64/egg/PIL/Jpeg2KImagePlugin.py to Jpeg2KImagePlugin.pyc
```

**Comment by wiredfool** [comment]
There should be lines like this where it's compiling the C code:
```
gcc -pthread -fno-strict-aliasing -g -O0 -Wall -Wstrict-prototypes -fPIC -DHAVE_LIBJPEG -DHAVE_OPENJPEG -DHAVE_LIBZ -DHAVE_LIBTIFF -I/usr/include/freetype2 -I/usr/local/include/openjpeg-2.0 -I/home/erics/Pillow/libImaging -I/home/erics/vpy27-dbg/include -I/usr/local/include -I/usr/include -I/usr/include/python2.7_d -I/usr/include/x86_64-linux-gnu -c libImaging/Jpeg2KDecode.c -o build/temp.linux-x86_64-2.7-pydebug/libImaging/Jpeg2KDecode.o
...
gcc -pthread -fno-strict-aliasing -g -O0 -Wall -Wstrict-prototypes -fPIC -DHAVE_LIBJPEG -DHAVE_OPENJPEG -DHAVE_LIBZ -DHAVE_LIBTIFF -I/usr/include/freetype2 -I/usr/local/include/openjpeg-2.0 -I/home/erics/Pillow/libImaging -I/home/erics/vpy27-dbg/include -I/usr/local/include -I/usr/include -I/usr/include/python2.7_d -I/usr/include/x86_64-linux-gnu -c libImaging/JpegEncode.c -o build/temp.linux-x86_64-2.7-pydebug/libImaging/JpegEncode.o
```

**Comment by wiredfool** [comment]
It doesn't look like grok is including `opj_config.h`, which is where the `OPJ_MAJOR_VERSION` is defined. (In openjpeg, it's included in `opj_stdint.h`)

**Comment by boxerab** [comment]
aha! Does Pillow inlcude opj_config.h, or opj_stdint.h ? Because I removed opj_stdint.h in Grok,
seeing as there is already a standard header for this.

**Comment by boxerab** [comment]
my guess is that Pillow is including `opj_stdint.h`, from a different location, with the wrong version information.

**Comment by boxerab** [comment]
since Grok doesn't have this header any longer.

**Comment by boxerab** [comment]
alrighty - I have added `opj_stdint.h` back into Grok. So,  Pillow should now be happy.

**Comment by wiredfool** [comment]
Nope, we're not doing it -- it's in openjpeg.h.

**Comment by boxerab** [comment]
Ahhh,now understand . Openjpeg.h is simply missing config.h

**Comment by boxerab** [comment]
No, opj_config.h *is* in openjpeg.h.  Well, with my latest checkin,  grok header files should be identical to openjpeg headers.

**Comment by boxerab** [comment]
correction: @wiredfool  you're right - config.h needs to be put in openjpeg.h

**Comment by boxerab** [comment]
@calvinbutcher I've updated the openjpeg.h header to include the config header file, so Pillow should now detect the correct version. I have also moved the `v2.1.1` tag to the latest commit, so Docker should be stable.

**Comment by calvinbutcher** [comment]
@boxerab  https://github.com/python-pillow/Pillow/issues/584#issuecomment-39160382 suggests that these self-test failures have been fixed by: 

> using openjpeg sources from `svn\tags\version.2.0` instead of `svn\branches\openjpeg-2.0`

Can I ask what tag/version of OpenJPEG was forked for Grok?

**Comment by calvinbutcher** [comment]
Sorry, for some reason 20 comments from 20 hours ago have just popped up on my screen. I'll catch up.

**Comment by calvinbutcher** [comment]
@wiredfool @boxerab that's awesome, thank you. I'll re-build and see what happens...

**Comment by calvinbutcher** [comment]
OK, Pillow 3.2.0 and Grok 1.0 now compile nicely together, which is great progress! :)

However, the Pillow nosetests still fail on the `make`, breaking the docker image build.

So, as I said before https://github.com/python-pillow/Pillow/issues/584#issuecomment-39160382 suggests that these self-test failures have been fixed by:

>using openjpeg sources from `svn\tags\version.2.0` instead of `svn\branches\openjpeg-2.0`

So the only thing I can think that can help here is to ask what tag/version of OpenJPEG was forked initially for Grok?

**Comment by boxerab** [comment]
Nice progress! Grok was forked from OpenJPEG about 6 months ago. This was the master branch, after the 2.1 release, but before the 2.1.1 release.  version-2.0 is quite old  - they haven't been using SVN in a long time.

Which version of openjpeg did you use for your Docker image, the one that built successfully ?

**Comment by boxerab** [comment]
Perhaps you can ask the Pillow folks to investigate Pillow building with the latest openjpeg release, which is `2.1.1`  ?

**Comment by calvinbutcher** [comment]
I'm checking out openJPEG `tags/version.2.1`. This has been successful with Loris/Pillow and IIPsrv.

I've not tried 2.1.1. Perhaps I should..

**Comment by boxerab** [comment]
Yes, my version is based on what became 2.1.1, and I think there may have been some breaking changes.  Also, you could open an issue with pillow about 2.1.1, if it doesn't work.

**Comment by calvinbutcher** [comment]
OK, I'll give it a try!

**Comment by calvinbutcher** [comment]
OK, the tag `version.2.1` is the latest official openjpeg release, except for `wg1n6848`? I'll try it with `master`.

**Comment by calvinbutcher** [comment]
Pillow 3.2.0 successfully installs against the `master` of openJPEG...

**Comment by calvinbutcher** [comment]
Yup, it passes.
```
Ran 702 tests in 17.348s

OK (SKIP=54)
check-manifest
lists of files in version control and sdist match
pyroma .
------------------------------
Checking .
Single threaded build, not installing mp_compile:1 processes
Found Pillow
------------------------------
Final rating: 10/10
Your cheese is so fresh most people think it's a cream: Mascarpone
------------------------------
viewdoc
```

**Comment by calvinbutcher** [comment]
Running the nosetests separately, all fine:

```
grok@grok-VirtualBox:~/pillow$ nosetests Tests/test_*.py
.....................SSSSSSS........................................................................SS...................S................................................S............S..............................................S..........................................................................................................SSSSSSSSSSSSS........................................................................S.................................................................SSSSS.............S......................................................................................S................................S..libpng warning: Invalid color type/bit depth combination in IHDR
libpng error: Invalid IHDR data
.S.................SSSSSS........SSSSSS..................SSSSSS..........
----------------------------------------------------------------------
Ran 702 tests in 19.068s

OK (SKIP=54)
```

**Comment by calvinbutcher** [comment]
I'll run the build again in case something has changed recently.

**Comment by calvinbutcher** [comment]
Actually, I think I've got my knickers in a twist here. Now Grok and Pillow compile together, I'm removing the manual Pillow compile. The standard python dist of Pillow doesn't do the tests, so fingers crossed...

**Comment by boxerab** [comment]
Sounds good.

**Comment by wiredfool** [comment]
If your tests are failing like the trace (way) above, then while it might have compiled, you're not going to be getting decent images out of it.

**Comment by calvinbutcher** [comment]
If by decent you mean clear, I'm actually getting clear images. My main objective was to get a docker image running with Loris and Grok. But the test issue, of course, still stands.

**Comment by calvinbutcher** [comment]
OK, we have a successful build for Loris 2.0.1 and Grok 1.0!

https://hub.docker.com/r/bdlss/loris-grok-docker/builds/

Thank you @boxerab @stweil @wiredfool and anyone else that I've forgotten :)))

**Comment by boxerab** [comment]
@wiredfool good point - I will look into the failing tests when I have time.  Thanks for your help.

**Comment by boxerab** [comment]
@wiredfool I think the AssertionError you mentioned was happening when there was a version mismatch due to missing config file.  @calvinbutcher do you still get this error ? Perhaps you are not running the tests?

**Comment by calvinbutcher** [comment]
@boxerab I'll do a manual compile and check it tomorrow. I'm pretty sure they were still failing but I'll confirm.

**Comment by boxerab** [comment]
cool. thanks.  By the way, are you planning on deploying these servers in production?

**Comment by calvinbutcher** [comment]
The remit I was given was to produce docker images (for Loris and IIPsrv w/ openjpeg and Grok) for ease of deployment/testing/demonstration etc. as part of a wider project/other initiatives. Why do you ask?

**Comment by boxerab** [comment]
Thanks. I'm keen to have people using the library, so interested in how it will be used in the wild.

**Comment by boxerab** [comment]
Feel free to close this issue if you think it has been resolved.

**Comment by calvinbutcher** [comment]
Thank you kindly for your help, Aaron. I'm just tidying things up but currently docker seems to be hanging. I'll be keep in touch with Grok's progress and no doubt we'll speak again! Take it easy! C

**Comment by boxerab** [comment]
My please. Keep it real.  
Aaron
